Update UiRequest.py

hacktivist opened the pull request:
Fonts are NOT applications! Added PGP related MIME types.

Commit:
Update MIME

Fonts are NOT applications! Added PGP related MIME types.

purplesyringa commented:
Why are the `.png` (and similar) changes required? `mimetypes` module sets correct mimetypes for (most) images already.

ghost commented:
Feel free to add or remove MIME types. I think it's better to have common extensions stated than rely on the guessing of the `mimetype` module. This is my preference, that is why I included the .png, .webm and others.

purplesyringa commented:
It's not *guessing*. The `mimetypes` module detects mimetype based on the extension, and `.png` is hardcoded. You can take a look at its source code if you're not sure.

ghost commented:
`else: content_type = mimetypes.guess_type(file_name)[0]` **is guessing** by checking the extension and then looking up in the module the supposedly correct MIME type. I think that ZeroNet should not rely on the module since some types are incorrect or not included at all like the PGP MIME types which I stated.

purplesyringa commented:
> since some types are incorrect

Which ones?

I see no reason in adding unneeded code when it works correctly. To be honest, I think that it's mimetypes module that should be patched, not ZeroNet code, but that's unlikely.

purplesyringa commented:
I'd leave `.js`, `.webp`, fonts and gpg stuff only. I'm also not sure whether `.js` is required (`application/javascript` is a correct mimetype as well but old browsers might not support it). Everything else is correct.

HelloZeroNet commented:
> Type name:
> application

According to https://www.iana.org/assignments/media-types/application/font-woff

Commit:
Update UiRequest.py

Remove some MIME types because mimetype module correctly provides them. text/plain, text/html and text/css is not removed. Fonts MIME types are fixed because they are not applications. I have added PGP MIME types for keys, encrypted and for signature. 
I declared also utf8 encoding for text/css, application/json and for application/javascript. 
JavaScript should be application/javascript and not text/javascript according to [RFC4329] (https://tools.ietf.org/html/rfc4329#page-9).

purplesyringa commented:
@HelloZeroNet [Take a look at this RFC](https://tools.ietf.org/html/rfc8081).

purplesyringa commented:
@HelloZeroNet @filips123 @krixano Do you see any problems in using `application/javascript` instead of `text/javascript`? I'd leave `text/javascript` but we could remove it (and thus simplify code) if `application/javascript` works as well.

purplesyringa commented:
`.txt`, `.html` and `.css` are handled by mimetypes module as well. BTW, I'd recommend not to use `.html` extension to detect mimetype because ZeroNet uses `text/html` mimetype to detect when to add sandbox. Mentioning `.html` might make it look like if `mimetypes` module could be removed, but there are other (strange) extensions like `.htm` and `.xht`, so removing might have serious consequences for ZeroNet security.

filips123 commented:
@imachug Regarding JS mime type, I already searched about this around the web and it seems that  nobody actually know which one is correct.

Also, if you are already adding new mime types, you should also add `.webmanifest` ==> `application/manifest+json` mime type as this is official mime type for Web App Manifest for PWA. It was already added to `mimetypes` module (by me), but the change is only present in Python 3.8+.

HelloZeroNet commented:
```
$ curl --head https://github.githubassets.com/static/fonts/inter/Inter-Medium.woff | grep content-type
Content-Type: application/font-woff
```

But if we want to change it, then
```
        elif ext in ("woff", "woff2", "ttf"):
            content_type = "font/%s" % ext
```
would be more clean/compact solution

ghost commented:
Yeah @imachug that's correct! https://www.iana.org/assignments/media-types/media-types.xhtml#font and [[RFC8081](https://tools.ietf.org/html/rfc8081)] states **`font/`** instead of `application/` MIME for fonts.

ghost commented:
@filips123 JavaScript should be `application/javascript` and not text/javascript according to [RFC4329](https://tools.ietf.org/html/rfc4329#page-9).

ghost commented:
This pr also belongs in the mimetypes repo - don't blame ZeroNet devs for things that aren't of their doing. In the meantime, a workaround is good, because this fonts issue has been a problem for a while I think.

HelloZeroNet commented:
> fonts issue has been a problem for a while I think.

As far I know there were no issue reported that is connected to invalid mime types.

All issues I have seen was because if someone enables the UiPassword plugin, then the cookie won't be passed to the css font requests and the client will display the login form instead of the resource.

ghost commented:
Idk, I'm pretty sure there was a comment in one issue that mentioned this problem with the mime type. Maybe I can find it, idk.

ghost commented:
@HelloZeroNet It's right here: #1940 
Tangdou closed it before it was even fixed. I made a comment on it recently.

ghost commented:
Hold on a sec... Previously, I wasn't getting correct results when running this in python when that issue (#1940 ) originally occured:
> print(__import__('mimetypes').guess_type('MaterialIcons-Regular.woff'))

But now I am getting "font/woff" in python. I suppose the mimetypes module was fixed for fonts but the zeronet workaround wasn't removed or something? idk...

Aso, @HelloZeroNet , some people were still getting font problems in Tor browser very recently... don't know if this pr fixes it though.

Commit:
Update UiRequest.py

Removed some MIME types because mimetypes module correctly provides them. text/plain, text/html and text/css is not removed. 
Font MIME types are fixed because they are not applications. All font MIME types from https://www.iana.org/assignments/media-types/media-types.xhtml#font is stated in a clean/compact fashion as @HelloZeroNet suggested https://github.com/HelloZeroNet/ZeroNet/pull/2248#issuecomment-546611519. 
I have added PGP MIME types for keys, encrypted and for signature. 
I declared also utf8 encoding for text/css, application/json and for application/javascript. 
JavaScript should be application/javascript and not text/javascript according to [RFC4329] (https://tools.ietf.org/html/rfc4329#page-9).
I don't add application/manifest+json at this time but you are free to do @filips123

HelloZeroNet commented:
> @HelloZeroNet It's right here: #1940
> Tangdou closed it before it was even fixed. I made a comment on it recently.

As he stated in last comment: "When enable the ui_password plugin , the problem will reproduce ."

So it was because of the sandbox.

filips123 commented:
> But now I am getting "font/woff" in python. I suppose the mimetypes module was fixed for fonts but the zeronet workaround wasn't removed or something?

Which Python version do you have? Because `mimetypes` is in standard library, it is updated with Python. So maybe this was changed in recent Python versions.

purplesyringa commented:
@filips123 @krixano `mimetypes` library has a small built-in list of common extensions, but other mimetypes are loaded from registry (in Windows) and configuration files (in Linux).

ghost commented:
@filips123 
> Why you didn't add Web App Manifest. If I would add it, this would be another PR and if you already have PR that updated mime types this would be unnecessary complicating.

See: https://github.com/w3c/manifest/issues/689#issuecomment-408558218

filips123 commented:
@imachug There are around 100 built-in extensions but yes other are loaded from system.

purplesyringa commented:
@hacktivist It looks like the comment you're referring too agrees with @filips123, doesn't it?

filips123 commented:
@hacktivist Can you explain that comment a bit more? Because I see it as point **for** using `.webmanifest` extension and correct content type.

ghost commented:
> As he stated in last comment: "When enable the ui_password plugin , the problem will reproduce ."
> 
> So it was because of the sandbox.

First of all, that last comment could be interpreted in two different ways... that the problem occurs with ui_password enabled or that it occurs with ui_password enabled **also**.

Secondly, don't take things out of context and actually look at the issue and what I quoted for my last comment... previously in the issue I tested the mimetypes library and it didn't give the correct results for fonts **in the python REPL** which indicated a potential mimetypes library or OS config problem. Hence why **you** suggested this:

> So it depends on the operating system configuration if its application/font-woff or None. In case of None we return "application/octet-stream" as content-type which works for browsers I tested.
> 
> To make it more predictable we could write our own content-type resolver that does not depends on operating system configuration files.

If operating systems are inconsistent, then that's a problem that obviously needs to be worked around. Also notice that you said you'll return "application/octet-sream" if None was given back...

ghost commented:
@filips123 
Quote from https://github.com/w3c/manifest/issues/689#issuecomment-408558218
> If a developer doesn't have access to server configurations then `.json` would be the safest way to deploy the manifest.

Also please look at [Web App Manifest #warning](https://w3c.github.io/manifest/#issue-container-generatedID)

purplesyringa commented:
Right, but we *do* have access to server configurations, don't we?

> The official file extension for the manifest is .webmanifest. Some web servers recognize this extension and transfer the file using the standardized media type for a manifest (application/manifest+json). Developers can also choose a different extension (e.g. .json) or none at all (e.g. /api/GetManifest), but are strongly encouraged to transfer the manifest using the application/manifest+json media type.

purplesyringa commented:
Sorry, my browser lagged so I didn't see you were referring to the warning. Anyways, mimetypes are usually rather stable so I doubt they'll change it.

ghost commented:
@imachug He can use `<link rel="manifest" href="manifest.webmanifest">` or `<link rel="manifest" href="manifest.json">` and the `mimetypes` module will do its job.

ghost commented:
Ok, well your previous comments suggested you thought it shouldn't be added because of webmanifest's instability/non-standardization...

purplesyringa commented:
@hacktivist Why use hacks when a small fix is enough?

ghost commented:
Should I add back `application/json`? I removed because `mimetypes` module states correctly. If we add `application/manifest+json` than I think `application/json` also needs to be added.

purplesyringa commented:
> If we add application/manifest+json than I think application/json also needs to be added.

Why though? I think that adding application/manifest+json is enough.

purplesyringa commented:
Do you mean that application/manifest+json is already reported correctly by mimetypes module? This only works on *some* devices because it's not built-in.

filips123 commented:
Mimetypes module only resolves .webmanifest in Python 3.8 or if this is in system configuration. But in most cases, it is not and most users don't have Python 3.8 yet.

And the only change needed for this is to check for `.webmanifest` and return mime type `application/manifest+json`. Changing `.json` is not needed.

ghost commented:
I adding both. Do we need also `application/javascript`? If html, css etc stated than we should add this one too. What about encoding? `.webmanifest` also should be utf8 right?

purplesyringa commented:
> I adding both.

But why? Why would you duplicate mimetypes's work?

> Do we need also application/javascript? If html, css etc stated than we should add this one too.

We should *remove* html and css.

> What about encoding? .webmanifest also should be utf8 right?

I think so.

ghost commented:
I adding `application/javascript` for `.js`, `application/json` for `.json` and `application/manifest+json` for `.webmanifest`. I will not remove `text/simple`,` text/html` and `text/css` because from line 303 these MIME types are explicitly stated with `utf8` encoding.

Commit:
Update UiRequest.py

See: https://github.com/HelloZeroNet/ZeroNet/pull/2248#issuecomment-546621653

purplesyringa commented:
< I adding application/javascript for .js, application/json for .json

But why? .js and .json are already handled by mimetypes.

> I will not remove text/simple

Does that make sense? I mean, these two parts are unrelated...

> text/html and text/css because from line 303 these MIME types are explicitly stated with utf8 encoding.

Ok, feel free to ignore these because the diff would be smaller.

HelloZeroNet commented:
> @krixano Also notice that you said you'll return "application/octet-sream" if None was given back...

Loading fonts with application/octet-sream content-type should work.

Eg. Cloudflare/Firefox serve it this way:

```
$ curl --head https://www.mozilla.org/media/fonts/Metropolis-Bold.6a80125e795a.woff2
Content-Type: binary/octet-stream
```

So I still think the problem was because of missing cookie header and not because of invalid content type, but including more content types in UiRequest.py should not hurt.

HelloZeroNet commented:
I prefer serving js/css/json as `text/*`, because this way if you open in a browser you can view the content instead of downloading it.

Google.com also serves it this way, so I think it can be considered a valid option.

Update: Looks like in latest chrome js/json also displayed and not downloaded when serving it as application/* (maybe it got changed or I was remember incorrectly)

ghost commented:
> > @krixano Also notice that you said you'll return "application/octet-sream" if None was given back...
> 
> Loading fonts with application/octet-sream content-type should work.
> 
> Eg. Cloudflare/Firefox serve it this way:
> 
> ```
> $ curl --head https://www.mozilla.org/media/fonts/Metropolis-Bold.6a80125e795a.woff2
> Content-Type: binary/octet-stream
> ```
> 
> So I still think the problem was because of missing cookie header and not because of invalid content type, but including more content types in UiRequest.py should not hurt.

It should be served as `font/` because if it's not than ZeroNet will use the `mimetypes` module which will serve it as `application/` which is wrong.

Do not even think about to follow or listen what CloudFlare does or say because than I will be forced to pull out all of my hair one-by-one.

ghost commented:
@imachug `application/manifest+json` is also handled by the `mimetypes` module https://github.com/python/cpython/blob/3.8/Lib/mimetypes.py at least in Python3 which should be the target of ZeroNet because January 1st 2020 will be Python2's EOL (End of Life).

filips123 commented:
@Hacktivist Only since Python 3.8. But most users don't use it yet.

purplesyringa commented:
@hacktivist 

> 3.8

> at least in Python3

Commit:
Compact :)

On my gateway I use for text/html the Link header also because I serving files (font, logo etc) using HTTP/2.0 with PUSH.

purplesyringa commented on src/Ui/UiRequest.py:
```diff
@@ -300,18 +300,8 @@ def sendHeader(self, status=200, content_type="text/html", noscript=False, allow
-            content_type = "application/json; charset=utf-8"
-        if content_type == "application/manifest+json":
-            content_type = "application/manifest+json; charset=utf-8"
+        if content_type in ("text/plain", "text/html", "text/css", "application/js", "application/json", "application/manifest+json"):
```
application/js -> application/javascript

hacktivist replied:
:O Thanks!

Commit:
Update UiRequest.py

ghost commented:
> Loading fonts with application/octet-sream content-type should work.
> 
> Eg. Cloudflare/Firefox serve it this way:
> 
> $ curl --head https://www.mozilla.org/media/fonts/Metropolis-Bold.6a80125e795a.woff2
Content-Type: binary/octet-stream
> 
> So I still think the problem was because of missing cookie header and not because of invalid content type, but including more content types in UiRequest.py should not hurt.

@HelloZeroNet I don't think you get what I was saying at all. You were trying to say that that other issue wasn't about this at all, but you literally propose a "fix" for this exact problem in that other issue (and you never actually stated that it was fixed) - so it **was** about this.

> So I still think the problem was because of missing cookie header and not because of invalid content type

Because apparently you can't have two problems at once right? I was having that content-type font problem also, and I wasn't using ui_password (and I probably have photo-evidence of this somewhere on ZeroMe)... so that goes to show.
I mean, logically, the problem was *very clearly* **both** ui_password and a wrong content-type on fonts considering both YOU and I tested it and it wasn't giving back the correct content-type, but let's just forget about that why don't we!

Anyways @HelloZeroNet , icon fonts don't currently work in Tor browser - fix your crap!

New Issue: #2249

Commit:
Reverse Tabnabbing mitigation in ZeroNet by sending Referrer-Policy: "no-referrer" header.

Adding Referrer Policy to mitigate Reverse Tabnabbing (https://www.owasp.org/index.php/Reverse_Tabnabbing)
See recent commit escaping sandbox using target _blank: https://github.com/HelloZeroNet/ZeroNet/commit/e82155aac4e75b1e67ae3f569a912b8f40fc8f48

ghost commented:
The MIME type `font/*` should be used on **fonts** because, I do serving ZeroNet content with a reverse proxy and some resource is cached on the client, more specifically the 1,5 MB `woff2`.

Commit:
Caching

Like this.

Commit:
Update UiRequest.py

Update MIME type from text/javascript to application/javascript. 
Note, that MIME type application/* is not caching. 
See: https://github.com/HelloZeroNet/ZeroNet/pull/2248/commits/10d44cf10812df6ac57e49c83532832527cdd2bc

Commit:
Update UiRequest.py

Let me know if this don't work.

Commit:
Update UiRequest.py

font-src '127.0.0.1:43110'

Commit:
Update UiRequest.py

*

ghost commented:
@imachug Now the fonts working? :)

purplesyringa commented on src/Ui/UiRequest.py:
```diff
@@ -285,24 +299,41 @@ def sendHeader(self, status=200, content_type="text/html", noscript=False, allow
+            
+        # UTF-8 character encoding
+        if content_type in ("text/plain", "text/html", "text/css", "application/javascript", "application/json", "application/manifest+json"):
+            content_type = "%s;" % content_type + " " + "charset=utf-8"
```
What magic is this??

hacktivist replied:
This magic works. If you now start arguing about UTF8 character encoding... What you don't like?

purplesyringa replied:
Formatting and concatenation in the same expression, lol.

hacktivist replied:
Anything is possible in Python. :rofl:

filips123 replied:
Just use `content_type = content_type + " " + "charset=utf-8;"`. This current version is really weird...

hacktivist replied:
**text/css; charset=utf-8** :heavy_check_mark: 
~~text/html charset=utf-8;~~

hacktivist replied:
**text/html; charset=utf-8** :heavy_check_mark: 
~~text/html charset=utf-8;~~

filips123 replied:
I misplaced `;`. I mean `content_type = content_type + "; charset=utf-8"` or just `content_type += "; charset=utf-8"`.

purplesyringa commented on src/Ui/UiRequest.py:
```diff
@@ -285,24 +299,41 @@ def sendHeader(self, status=200, content_type="text/html", noscript=False, allow
-            headers["Cache-Control"] = "public, max-age=600"  # Cache 10 min
-        else:
-            headers["Cache-Control"] = "no-cache, no-store, private, must-revalidate, max-age=0"  # No caching at all
+        if status in (200, 206) and cache_txt_css_html:  # Cache MIME type text/* for 1 week
```
AFAIK we don't do cache invalidation so this might be a problem. Static site update might take a long while...

hacktivist replied:
Cache is `public` as you see. In the other hand after `siteSign` there should be `sitePublish` which will notify all peers about the changes.

purplesyringa replied:
How does this help? I mean, the browser would still use cached html, wouldn't it?

hacktivist replied:
Yes, but not for 10 minutes. It would be crazy to assume that a static sites changes in every 10 minutes. 
Does the browser cache `content.json`? No, is not because we are not caching any MIME type **application/***, specially not JavaScript.

purplesyringa replied:
It's crazy to assume 10 minutes but it's crazy to assume a whole week too. I'd recommend 2 hours or so.

purplesyringa replied:
How is content.json related? It's not accessed by browser at all.

purplesyringa replied:
Ok, it looks like you thought of site updating as of publishing. I actually meant that users would receive updates but their browsers wouldn't show the changes.

hacktivist replied:
Note:
On line 299 in `cacheable_type` JavaScript was `application/javascript`
`            self.env["REQUEST_METHOD"] == "OPTIONS" or content_type == "application/javascript"`

On line 272 which was also [removed](https://github.com/HelloZeroNet/ZeroNet/pull/2248/commits/091f30fbd7163a28736e970f9a432b67d4dab4c7), JavaScript used as `text/javascript`
`        if content_type == "text/javascript" and not self.env.get("HTTP_REFERER"):`

hacktivist replied:
> Ok, it looks like you thought of site updating as of publishing. I actually meant that users would receive updates but their browsers wouldn't show the changes.

Depend, I seen that many people rocking with Tor Browser. :man_technologist: 
Caching is handy when you have a gateway or ZeroNet is behind a reverse proxy.

purplesyringa replied:
What about making a `--cache_for` configuration option then? @HelloZeroNet

hacktivist replied:
I can add `stale-while-revalidate` to MIME type **text/*** if that helps.

purplesyringa replied:
Ok, I guess that makes sense. I'd still add `--cache_for` though, but I guess it should be another PR.

hacktivist replied:
@imachug I don't think that need `--cache_for`. 
I do have 200+ static websites and I can tell you for sure that I'm not updating them every 10 minutes. Not even in one day. Caching is good how you see in these changes.

purplesyringa replied:
In case of ZeroNet, static websites are often actually made with jekyll or similar tools (look at @caryoscelus'es blog). *These* kinds of sites are updated rather often.

hacktivist replied:
Nothing is stopping him in lowering the default `max-age` value if needed.

purplesyringa replied:
Who is "him"? @caryoscelus?

hacktivist replied:
Yes. You come up with one specific example and I replied accordingly. 

According to my checks by refreshing the site in the browser you will eventually get a fresh copy of the static resources. 
When you use _back_ and _forward_ button than obviously you will see the cached version. 

See in ZeroNet `frame` and using `raw`.

purplesyringa replied:
> Yes. You come up with one specific example and I replied accordingly.

caryoscelus (or whoever runs a site) can't just change caching settings on *others'* computers. How would you even do that?

> According to my checks by refreshing the site in the browser you will eventually get a fresh copy of the static resources.

This "eventually" often includes accidentally pressing Shift / Ctrl or doing something in web console. That's not something that an average user should think about.

> See in ZeroNet frame and using raw.

Now what?.. Are you asking me to use raw mode? That's often not an option -- static sites *can* include JS, e.g. for animations.

hacktivist replied:
Let's say you have a chat in ZeroNet and you send `text/html` to the cache for one week while using JavaScript to print messages. Not to mention you can use JavaScript to refresh the site time-to-time without needing to mess around the default caching of the resources. Correct me if I'm wrong.

purplesyringa replied:
Let's talk about a specific example -- a static blog. Pretend that I had this blog on clearnet (e.g. on GitHub pages) but then I moved it to ZeroNet. So, my blog doesn't work with ZeroNet API, it's just a collection of html files and a few JS files that are used for animations and other interactive content. Now, how would you reload site content when a new post is published?

hacktivist replied:
No MIME type starts with `application/*` is cached! This includes JavaScript, JSON an others like `application/octet-stream`. 
You can use query string on static resources like ZeroNet does, example: **all.css?rev=666**.

purplesyringa replied:
We're talking about *HTML pages*, not scripts!

hacktivist replied:
Let me be very clear on that ZeroNet is not in development in order to provide hosting for Jekyll! If you want a blog than there are a tons of options. You can **clone ZeroTalk** for example or just go and run it on WordPress. 
Most people don't even use this frameworks and it will be fatal to assume that HTML pages must work like a dynamic blog! I say it again. The proposed caching is good.

purplesyringa replied:
What the heck? ZeroNet aims at easy single-click zite creation. It should be easy to both create sites from scratch and copy clearnet sites. You can't just say "we won't do that"!

hacktivist replied:
I can also tell you this in Hungarian if you need. I born in Hungary, but left the country when Orban come into power nearly 10 years ago. My sister is a "border hunter" in the Croatian border, hunting real living people. I strongly believe that ZeroNet can really hurt those in power and I will not going to argue with you about how long the HTML should be cached. 

I, for that matter will always use this caching because as I said I do have 200+ static websites. It is already a pain to sign the new content and I don't want to see my server overloaded with request for static resources like HTML, CSS, video and other which are currently fetched every 10 minutes.

purplesyringa replied:
Ok, that's why I asked to add a `--cache_for` option. It could be set to 2 hours by default and increased by those who have lots of sites.

krixano replied:
> Most people don't even use this frameworks and it will be fatal to assume that HTML pages must work like a dynamic blog! I say it again. The proposed caching is good.

Yeah, you're an idiot.

hacktivist replied:
Choose your words you retard!

hacktivist replied:
I will send another PR but the caching stays how it is. Nobody and I say it again NOBODY in this word changes HTML pages every 10 minutes or so. You can use JavaScript or whatever you want if you need to update the text. Show me an example HTML site where the text changes every 10 minutes otherwise be silent.

krixano replied:
I'm not the one here who thinks that everything *has* to use javascript to load content.

And you know what ZeroNet **is** in development to do... provide websites to users! LOL Jekyll is a generator for a website, right?

Finally, Jekyll isn't what he was talking about - Jekyll loads things dynamically via javascript I believe (I might be wrong about this).

hacktivist replied:
I will send another PR but the caching stays how it is. Nobody and I say it again NOBODY in this world changes HTML pages every 10 minutes or so. You can use JavaScript or whatever you want if you need to update the text. Show me an example HTML site where the text changes every 10 minutes otherwise be silent.

hacktivist replied:
Go fckyourself with you Jekyll.

krixano replied:
If you think that static zites aren't updated more than every week, then yeah... you're an idiot. If you think that zites have control over caching of stuff, then you're just assuming ZeroNet has this features without ever seeing it anywhere.

hacktivist replied:
You know what I will not send a PR because you are a disgusting idiot who probably never in his life wrote a website from 0 with HTML. You just talk shit about programs that generate HTML sites but you fail to come up with an example where HTML is changes every 10 minutes. 

ZeroNet should check the content.json and if the hash is different than refresh the site. That would be the solution but instead of writing the code up you are here to spit shit.

krixano replied:
> ZeroNet should check the content.json and if the hash is different than refresh the site. 

You clearly have no idea what you're talking about. LOL

krixano replied:
> sites but you fail to come up with an example where HTML is changes every 10 minutes.

You know how I can tell you're a **very fishy** person... because you keep spewing that everyone's saying "10 minutes" when nobody has said that. Everyone here has only said changes happen shorter than **a week**.

I'll be making my own PR.

Btw, I **did** have a static zite, lol. And it changed frequently.

hacktivist replied:
@krixano Seriously, you opened a fucking issue about font loading and I did fixed it. You are an idiot and you know it because you was unable to fix such a simple thing as font loading in Tor Browser.

hacktivist replied:
You make your own PR but copy nothing from mine!

krixano replied:
Right... Because apparently everyone who opens an issue also can't program? Because apparently people you think that there's **no other** reason for not fixing the issue themselves. Because apparently people don't have work or school or anything else going on in their life. Again a **very fishy person**.

Btw, there's a different between fixing something and half-assing it (and deliberately preventing things from work because you're an asshole).

hacktivist replied:
I think you should be kicked out of this repo because you appreciate nothing but talking shit. ZeroNet should really distance itself from people like you. You ruining it not developing. What developing, you don't know even how to fix font loading or where to look. You are a disaster and you look like one according to your photo, website etc..

krixano replied:
Right... says the person who also said crap about ivanq who is the second highest for most commits and has fixed more vulnerabilities than you've even found in ZeroNet, LOL

hacktivist replied:
I'm sorry? I come to this repo 2 days ago! 
Sorry that there is no a million PR already. How can be someone so retarded like you?

krixano replied:
> You make your own PR but copy nothing from mine!

Does this mean you're claiming copyright for your commit?

@HelloZeroNet I'd be weary of the potential legal consequences of this in the future. This is precisely why open-source repos have a file that requires all committers to give up rights to their committs.

hacktivist replied:
I'm Hungarian too just that you know.

Commit:
Update UiRequest.py

Commit:
Update UiRequest.py

Looks way better.

Commit:
Update UiRequest.py

https://github.com/HelloZeroNet/ZeroNet/pull/2248/commits/5966d713230bc7b0d0a34e4907380199b6c699f4#r339349062

ghost commented:
:disappointed: I sent the wrong link in https://github.com/HelloZeroNet/ZeroNet/pull/2248/commits/99bea958e79ebcfff5669df6c724193d70ed3140 that should be: https://github.com/HelloZeroNet/ZeroNet/pull/2248#discussion_r339349062

HelloZeroNet commented:
I have pushed modifications in Rev4250 that has updated list of bundled content types based on this PR and also solves the font loading problems.
Adding long cache values is not a great idea as the content can change any time.

ghost commented:
Also posted on https://github.com/antifa/ZeroNet/issues/1

Here's a 3 way diff. Center is the original before hacktivist/antifa's PR. Left is nofish's implementation. Right is hacktivist/antifa's implementation from the PR (https://github.com/HelloZeroNet/ZeroNet/pull/2248)

![Screenshot from 2019-11-05 00-28-20](https://user-images.githubusercontent.com/4098042/68183936-600fbc80-ff63-11e9-9e9d-a738de560c57.png)
![Screenshot from 2019-11-05 00-27-28](https://user-images.githubusercontent.com/4098042/68183943-643bda00-ff63-11e9-85ee-6a1d8495bf04.png)
![Screenshot from 2019-11-05 00-27-47](https://user-images.githubusercontent.com/4098042/68183947-66059d80-ff63-11e9-88fe-49786fedd3e3.png)
![Screenshot from 2019-11-05 00-28-03](https://user-images.githubusercontent.com/4098042/68183951-67cf6100-ff63-11e9-853d-33238396e591.png)

Here's a video of me scrolling through the whole file:
https://www.youtube.com/watch?v=YaxcDekczYo

Here's another diff I did - a two way diff comparing ZeroNet current code with antifa/hacktivist's PR:
https://www.diffchecker.com/NO4OAFuM
(Note: because this is two way, it doesn't show which parts were in there originally before the PR or nofish's implementation).

ghost commented:
I'm posting here to update people on what's going on with the user @hacktivist / @antifa. After antifa was blocked, he created a new account called @LiberateZeroNet. He has deleted his account again and created a new account called @fukcyouintheass. This account has already been deleted, and 6 hours ago he has created *another* new account called AnonymousChronicler. I believe that he has an additional account that he has created an hour ago called bela-miklos. Both of these accounts have one repository that happens to be a copy of ZeroNet source code (but not shown as a fork, and under a different name).

Every single time I find out about an additional username this person created, I report it to GitHub Support where I have an issue open with them about the user hacktivist all the way back from when he still used that name - I created this report with GitHub Support on October 30th.

![Screenshot from 2019-11-05 19-02-08](https://user-images.githubusercontent.com/4098042/68259251-fe049500-fffe-11e9-9ce8-4bd1577e256b.png)
![Screenshot from 2019-11-05 19-02-25](https://user-images.githubusercontent.com/4098042/68259253-ffce5880-fffe-11e9-9607-256a7af36e6e.png)
![Screenshot from 2019-11-05 19-02-40](https://user-images.githubusercontent.com/4098042/68259256-00ff8580-ffff-11e9-9f3c-ec375f9a4b27.png)
![Screenshot from 2019-11-05 19-02-55](https://user-images.githubusercontent.com/4098042/68259259-02c94900-ffff-11e9-8f23-50aa781cc307.png)
![Screenshot from 2019-11-05 19-03-13](https://user-images.githubusercontent.com/4098042/68259264-03fa7600-ffff-11e9-8130-821710bef488.png)

ghost commented:
Another update - Looks like another account by this person has been created: PICSANRUGLAK
Joined an hour ago.
